services/galley/test/integration/API.hs

```diff
@@ -1239,6 +1253,146 @@ getConvIdsFailMaxSize = do
+              FederatedGalley.cmuUsersAdd = [(qAlice, roleNameWireMember)],
+              FederatedGalley.cmuUsersRemove = []
+            }
+    FederatedGalley.updateConversationMemberships fedGalleyClient cmu
+
+  -- 1 self conv + 2 convs with bob and eve + 197 local convs + 25 convs on
+  -- chad.example.com + 31 on dee.example = 256 convs. Getting them 16 at a time
+  -- should get all them in 16 times.
+  foldM_ (getChunkedConvs 16 alice) Nothing [15, 14 .. 0 :: Int]
+
+-- This test exists
+paginateConvListIdsPageEndingAtLocalsAndDomain :: TestM ()
+paginateConvListIdsPageEndingAtLocalsAndDomain = do
+  [alice, bob, eve] <- randomUsers 3
+  connectUsers alice (singleton bob)
+  connectUsers alice (singleton eve)
+  localDomain <- viewFederationDomain
+  let qAlice = Qualified alice localDomain
+  now <- liftIO getCurrentTime
+  fedGalleyClient <- view tsFedGalleyClient
+
+  -- With page size 16, 29 group convs + 2 one-to-one convs + 1 self conv, we
+  -- get 32 convs. The 2nd page should end here.
+  replicateM_ 29 $
+    postConv alice [bob, eve] (Just "gossip") [] Nothing Nothing
+      !!! const 201 === statusCode
+
+  -- We should be able to page through current state in 2 pages exactly
+  foldM_ (getChunkedConvs 16 alice) Nothing [1, 0 :: Int]
+
+  remoteChad <- randomId
+  let chadDomain = Domain "chad.example.com"
+      qChad = Qualified remoteChad chadDomain
+  -- The 3rd page will end with this domain
+  replicateM_ 16 $ do
+    conv <- randomId
+    let cmu =
+          FederatedGalley.ConversationMemberUpdate
+            { FederatedGalley.cmuTime = now,
+              FederatedGalley.cmuOrigUserId = qChad,
+              FederatedGalley.cmuConvId = Qualified conv chadDomain,
+              FederatedGalley.cmuAlreadyPresentUsers = [],
+              FederatedGalley.cmuUsersAdd = [(qAlice, roleNameWireMember)],
+              FederatedGalley.cmuUsersRemove = []
+            }
+    FederatedGalley.updateConversationMemberships fedGalleyClient cmu
+
+  remoteDee <- randomId
+  let deeDomain = Domain "dee.example.com"
+      qDee = Qualified remoteDee deeDomain
+  -- The 4th and last page will end with this domain
+  replicateM_ 16 $ do
+    conv <- randomId
+    let cmu =
+          FederatedGalley.ConversationMemberUpdate
+            { FederatedGalley.cmuTime = now,
+              FederatedGalley.cmuOrigUserId = qDee,
+              FederatedGalley.cmuConvId = Qualified conv deeDomain,
+              FederatedGalley.cmuAlreadyPresentUsers = [],
+              FederatedGalley.cmuUsersAdd = [(qAlice, roleNameWireMember)],
+              FederatedGalley.cmuUsersRemove = []
+            }
+    FederatedGalley.updateConversationMemberships fedGalleyClient cmu
+
+  foldM_ (getChunkedConvs 16 alice) Nothing [3, 2, 1, 0 :: Int]
+
+getChunkedConvs :: (Typeable a, FromJSON a) => Int32 -> UserId -> Maybe (Qualified ConvId) -> Int -> TestM a
+getChunkedConvs size alice start n = do
```

Comment:
I realize this is not for the `API.Util` module, but it also is not a stand-alone test. How about adding a short documentation for the function to make its purpose clear?

Reply:
Good idea, it took me a few minutes to understand what it did, I should document that :+1:

Reply:
This is done now, please let me know if it looks good.